add functors for standard and blaze math functions

moxcodes opened the pull request:
## Proposed changes

Put into `Utilities/Functional.hpp` are macros for unary, binary, and binary assign operators such
that functors can easily be generated from a function argument. This is then used to define
functors associated with all supported operations on `DataVectors`. For convenience in writing type-generic code, step_function, invsqrt, and invcbrt are defined for primitives.

### Types of changes:

- [ ] New feature

### Component:

- [ ] Code

### Code review checklist

- [ ] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`. For
  instructions on how to perform the CI checks locally refer to the [Dev guide
  on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run `make doc`
  to generate the documentation locally into `BUILD_DIR/docs/html`. Then open
  `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

nilsdeppe commented on src/Utilities/Functional.hpp:
```diff
@@ -0,0 +1,255 @@
+
+#include <complex>
+
+#pragma once
```
We've always had the pragma before includes

nilsdeppe commented on src/Utilities/Functional.hpp:
```diff
@@ -0,0 +1,255 @@
+ * supplied, e.g. `STRUCTNAME<>{}(val)`
+ *
+ */
+#define MAKE_UNARY_STRUCT(STRUCTNAME, FUNC)               \
```
I don't think we should document this. It should be an implementation detail. It also needs to be `#undef`ed at the end of the hpp, same with the binary one.

nilsdeppe commented on src/Utilities/Functional.hpp:
```diff
@@ -0,0 +1,255 @@
+ * \brief defines the inverse cube-root (\f$1/{}^3\sqrt{x}\f$) for arithmetic
+ * types
+ */
+template <typename T, Requires<std::is_arithmetic<T>::value> = nullptr>
```
Why do we need this requires? It's not used for SFINAE, so just remove it.

moxcodes replied:
I want to avoid defining the template for any types apart from arithmetic types for safety (e.g. regardless of how screwed up an include set is, this shouldn't run on a vector), because other definitions with the same name exist. Is there a better way to enforce that?

nilsdeppe replied:
The compilation will fail because the wrapped function doesn't exist for other types, so the desired behavior is already present :)

moxcodes replied:
true for step_function, not true for invsqrt. The sqrt function is defined for other types, so the interior of the function should work fine for DataVectors and (might?) shadow the blaze version, which is doubtlessly better optimized than just evaluating 1.0/sqrt(x).

nilsdeppe commented on src/Utilities/Functional.hpp:
```diff
@@ -0,0 +1,255 @@
+ * types and complex
+ */
+template <typename T,
+          Requires<std::is_arithmetic<T>::value ||
```
We use `or`,`and` and `not` instead of `||`, `&&` and `!`

nilsdeppe commented on src/Utilities/Functional.hpp:
```diff
@@ -0,0 +1,255 @@
+ */
+template <typename T,
+          Requires<std::is_arithmetic<T>::value ||
+                   cpp17::is_same_v<T, std::complex<double> > > = nullptr>
```
Why this requires? It's not sure for SFINAE so just removed it

nilsdeppe commented on src/Utilities/Functional.hpp:
```diff
@@ -0,0 +1,255 @@
+ * \brief defines the step function for arithmetic types
+ */
+template <typename T, Requires<std::is_arithmetic<T>::value> = nullptr>
+constexpr inline T step_function(const T& arg) {
```
This shouldn't be in functional since it's not a function object.

nilsdeppe commented on src/Utilities/Functional.hpp:
```diff
@@ -0,0 +1,255 @@
+ * \ingroup UtilitiesGroup
+ * \brief defines the step function for arithmetic types
+ */
+template <typename T, Requires<std::is_arithmetic<T>::value> = nullptr>
```
Unnecessary requires

nilsdeppe commented on src/Utilities/Functional.hpp:
```diff
@@ -0,0 +1,255 @@
+ *
+ */
+#define MAKE_BINARY_STRUCT(STRUCTNAME, FUNC)          \
+  template <typename T1 = void, typename T2 = void>   \
```
Having two template parameters is inconsistent with the STL: https://en.cppreference.com/w/cpp/utility/functional/plus

moxcodes replied:
This is an interesting point -- I think it is more desirable to have separate template parameters, and perhaps so much so to deviate from the STL. For my purposes in developing general tests, it would be nice to have a functor that works for testing the addition of, for instance, DataVectors and doubles, which is permitted for the raw operator +. If we want to deviate, though, we'd want our own Plus, Minus, etc. as well.

nilsdeppe commented on src/Utilities/Functional.hpp:
```diff
@@ -0,0 +1,255 @@
+  template <typename T1 = void, typename T2 = void>   \
+  struct STRUCTNAME {                                 \
+    auto operator()(const T1& arg1, const T2& arg2) { \
+      return (FUNC(arg1, arg2));                      \
```
Do we need the extra parens here? (I think clang-tidy or something sometimes wants them, so if that's the case that's fine)

nilsdeppe commented on src/Utilities/Functional.hpp:
```diff
@@ -0,0 +1,255 @@
+ *
+ */
+#define MAKE_BINARY_INPLACE_STRUCT(STRUCTNAME, FUNC) \
+  template <typename T1 = void, typename T2 = void>  \
```
These are potentially quite inefficient and don't actually do as advertised. They should actually call `+=`, etc.

nilsdeppe commented on src/Utilities/Functional.hpp:
```diff
@@ -0,0 +1,255 @@
+
+/*!
+ * \ingroup UtilitiesGroup
+ * \brief A struct wrapper for the abs function */
```
We've always put the closing `*/` on its own line everywhere else in the code

nilsdeppe commented on src/Utilities/Functional.hpp:
```diff
@@ -0,0 +1,255 @@
+    auto operator()(const T& arg) {                       \
+      return (FUNC(arg));                                 \
+    }                                                     \
+  };
```
Remove `;` from end otherwise we will get warnings on some compiler setups.

nilsdeppe commented on tests/Unit/Utilities/Test_Functional.cpp:
```diff
@@ -0,0 +1,232 @@
+MAKE_BINARY_INPLACE_TEST(multassign, MultAssign<>{}, std::multiplies<>{});
+MAKE_BINARY_INPLACE_TEST(plusassign, PlusAssign<>{}, std::plus<>{});
+
+using RealTypeList = brigand::list<float, double>;
```
Why are you using `brigand`? This isn't done anywhere else, you should use `tmpl`

nilsdeppe commented on tests/Unit/Utilities/Test_Functional.cpp:
```diff
@@ -0,0 +1,232 @@
+MAKE_BINARY_TEST(ATan2<>{}, atan2)
+MAKE_BINARY_TEST(Hypot<>{}, hypot)
+
+MAKE_BINARY_INPLACE_TEST(divassign, DivAssign<>{}, std::divides<>{});
```
`div_assign` instead of `divassign`, similar to rist

nilsdeppe commented on tests/Unit/Utilities/Test_Functional.cpp:
```diff
@@ -0,0 +1,232 @@
+MAKE_UNARY_TEST(ACosh<>{}, acosh)
+MAKE_UNARY_TEST(ASin<>{}, asin)
+MAKE_UNARY_TEST(ASinh<>{}, asinh)
+MAKE_UNARY_TEST(ATan<>{}, atan)
```
`ATan` -> `Atan`, similar to all the others everywhere that do the current thing.

nilsdeppe commented on tests/Unit/Utilities/Test_Functional.cpp:
```diff
@@ -0,0 +1,232 @@
+
+  brigand::for_each<RealTypeList>(
+      [&generator, &generic_bounds, &positive_bounds](auto x) {
+        test_func_eval_cbrt<typename decltype(x)::type>(
```
This might be a good place to use `BOOST_PP_LIST_FOR_EACH`. `GENERATE_INSTANTIATIONS` does this and it saves a lot of repetition.

nilsdeppe commented on tests/Utilities/MakeWithRandomValues.hpp:
```diff
@@ -70,6 +70,53 @@ struct FillWithRandomValuesImpl<Variables<tmpl::list<Tags...>>,
+  using result_type = T;
+  std::array<T, 2> bounds;
+  template <typename Bound>
+  uniform_custom_distribution(Bound Low, Bound High) noexcept {
```
I think `Bound` should be `T`

nilsdeppe replied:
Use list initialization, i.e. `bounds{{low, high}}`

nilsdeppe replied:
`Low` -> `low` and `High` -> `high` (snake_case)

nilsdeppe commented on tests/Utilities/MakeWithRandomValues.hpp:
```diff
@@ -70,6 +70,53 @@ struct FillWithRandomValuesImpl<Variables<tmpl::list<Tags...>>,
+  std::array<T, 2> bounds;
+  template <typename Bound>
+  uniform_custom_distribution(Bound Low, Bound High) noexcept {
+    bounds = std::array<T, 2>{{T(Low), T(High)}};
```
No C-style casts

nilsdeppe commented on tests/Utilities/MakeWithRandomValues.hpp:
```diff
@@ -70,6 +70,53 @@ struct FillWithRandomValuesImpl<Variables<tmpl::list<Tags...>>,
+/// the generated type can be constructed. The default generation function
+/// operates as would be expected for a double
+template <typename T>
+struct uniform_custom_distribution {
```
Classes are CamelCase

nilsdeppe commented on tests/Utilities/MakeWithRandomValues.hpp:
```diff
@@ -70,6 +70,53 @@ struct FillWithRandomValuesImpl<Variables<tmpl::list<Tags...>>,
+/// operates as would be expected for a double
+template <typename T>
+struct uniform_custom_distribution {
+  using result_type = T;
```
`value_type` seems more consistent with the STL, why `result_type`?

moxcodes replied:
needs to be that way to interact properly with anything that uses the distribution as a distribution. Patterned after std::uniform_real_distribution [cpp reference - uniform_real_distribution](https://en.cppreference.com/w/cpp/numeric/random/uniform_real_distribution)

nilsdeppe commented on tests/Utilities/MakeWithRandomValues.hpp:
```diff
@@ -70,6 +70,53 @@ struct FillWithRandomValuesImpl<Variables<tmpl::list<Tags...>>,
+template <typename T>
+struct uniform_custom_distribution {
+  using result_type = T;
+  std::array<T, 2> bounds;
```
Does this need to be public?

nilsdeppe commented on tests/Utilities/MakeWithRandomValues.hpp:
```diff
@@ -70,6 +70,53 @@ struct FillWithRandomValuesImpl<Variables<tmpl::list<Tags...>>,
+/// components of a std:: complex<double
+template <>
+template <class Generator>
+std::complex<double> uniform_custom_distribution<std::complex<double>>::
```
Why only complex doubles and not all complex?

moxcodes replied:
That constitutes a partial specialization and requires a full class definition, rather than getting to just define the member function. This seemed briefer, but if you want the generality I can make a partial specialization.

nilsdeppe commented on tests/Utilities/MakeWithRandomValues.hpp:
```diff
@@ -70,6 +70,53 @@ struct FillWithRandomValuesImpl<Variables<tmpl::list<Tags...>>,
+
+  template <class Generator>
+  T operator()(Generator& g) noexcept {
+    return ((T(g()) - Generator::min()) *
```
No C-style cast

nilsdeppe commented on tests/Utilities/MakeWithRandomValues.hpp:
```diff
@@ -70,6 +70,53 @@ struct FillWithRandomValuesImpl<Variables<tmpl::list<Tags...>>,
+
+  template <class Generator>
+  T operator()(Generator& g) noexcept {
+    return ((T(g()) - Generator::min()) *
```
Why all this min() and max()?

moxcodes replied:
It's part of the specification for distributions given a random bit generator.
[cppreference - UniformBitGenerator](https://en.cppreference.com/w/cpp/named_req/UniformRandomBitGenerator)

nilsdeppe commented on tests/Utilities/MakeWithRandomValues.hpp:
```diff
@@ -70,6 +70,53 @@ struct FillWithRandomValuesImpl<Variables<tmpl::list<Tags...>>,
+/// the generated type can be constructed. The default generation function
+/// operates as would be expected for a double
+template <typename T>
+struct uniform_custom_distribution {
```
How is this better than fill_with_random_values?

moxcodes replied:
it's a level lower - fill_with_random_values takes a generator and a distribution as arguments, so you need a distribution to provide. And the std:: distributions are unsatisfactory for generating complex numbers as well as other types

nilsdeppe commented on tests/Unit/Utilities/Test_Functional.cpp:
```diff
@@ -0,0 +1,232 @@
+
+using RealTypeList = brigand::list<float, double>;
+
+using AllTypeList = brigand::list<float, double, std::complex<double>>;
```
metavariables -> `all_type_list`

nilsdeppe commented on src/Utilities/Functional.hpp:
```diff
@@ -0,0 +1,255 @@
+// Distributed under the MIT License.
```
Lots of missing `noexcept`

nilsdeppe commented on src/Utilities/Functional.hpp:
```diff
@@ -0,0 +1,255 @@
+// See LICENSE.txt for details.
+
+/// \file
+/// Defines function objects
```
Just remove the file comment. It's not required and I don't think it's useful enough here.

nilsdeppe commented on src/Utilities/Functional.hpp:
```diff
@@ -0,0 +1,255 @@
+/// \file
+/// Defines function objects
+
+#include <complex>
```
you're missing a lot of headers

nilsdeppe commented on tests/Unit/Utilities/Test_Functional.cpp:
```diff
@@ -0,0 +1,232 @@
+    CHECK(val1 == result);                                          \
+  }
+
+MAKE_UNARY_TEST(Abs<>{}, abs)
```
`BOOST_PP_LIST_FOR_EACH` is your friend

moxcodes replied:
Having looked it up a bit, I admire the madness of doing loops in macros, and I do see where it can be useful for generating a bunch of this type of thing, but because we have different tokens for the structure and the function it needs to compare against, a pairwise list of them needs to be somewhere. I suspect that the process of making that list and the extra macro syntax necessary to unpack the pairs ends up being sufficiently complicated to use and maintain that the listing out here is no worse. They scale the same, anyway =P

nilsdeppe commented on tests/Utilities/MakeWithRandomValues.hpp:
```diff
@@ -70,6 +70,52 @@ struct FillWithRandomValuesImpl<Variables<tmpl::list<Tags...>>,
+
+/// \ingroup TestingFrameworkGroup
+/// \brief Specialization of the generation function to independently sample the
+/// components of a std:: complex<double
```
why the space after`::`?

moxcodes commented:
Alright, this has been fully updated to new functionals state now.

nilsdeppe commented on src/Utilities/Math.hpp:
```diff
@@ -45,3 +47,28 @@ T evaluate_polynomial(const std::vector<U>& coeffs, const T& x) noexcept {
+/// \ingroup UtilitiesGroup
+/// \brief defines the step function for arithmetic types
+template <typename T, Requires<std::is_arithmetic<T>::value> = nullptr>
+constexpr inline T step_function(const T& arg) noexcept {
```
You don't need to mark these as inline, templates already form weak symbol for linking purposes.

You could force inline them, if you want. However, given how simple they are I expect the compiler to handle it.

nilsdeppe commented on src/Utilities/Math.hpp:
```diff
@@ -45,3 +47,28 @@ T evaluate_polynomial(const std::vector<U>& coeffs, const T& x) noexcept {
+/// and complex types
+template <typename T,
+          Requires<std::is_arithmetic<T>::value or
+                   cpp17::is_same_v<T, std::complex<double>>> = nullptr>
```
we have `tt::is_a_v` so you could cover all `std::complex` types.

nilsdeppe commented on docs/Doxyfile.in:
```diff
@@ -156,7 +156,10 @@ ENABLE_PREPROCESSING   = YES
 
 # Macros we actually want to expand during preprocessing
-EXPAND_AS_DEFINED      = MAKE_BINARY_OPERATOR
+EXPAND_AS_DEFINED      = MAKE_BINARY_OPERATOR \
```
We should have these be in alphabetical order otherwise it'll become out of control in the long run.

nilsdeppe commented on src/Utilities/Functional.hpp:
```diff
@@ -29,39 +60,120 @@ struct Functional {
-/// Functional for dividing two objects
-MAKE_BINARY_OPERATOR(Divides, /)
+#define MAKE_UNARY_FUNCTIONAL(NAME, OPERATOR)             \
+  /** Functional for computing OPERATOR on an object */   \
```
Is everything still working if you put `OPERATOR` inside ticks here and in the other macros?

nilsdeppe commented on src/Utilities/Functional.hpp:
```diff
@@ -121,5 +249,8 @@ struct Square : Functional<C::arity> {
   }
 };
 
+#undef MAKE_UNARY_FUNCTIONAL
```
Alphabet

nilsdeppe commented on src/Utilities/Functional.hpp:
```diff
@@ -7,11 +7,42 @@
+ *
+ * \details
+ * These chaining function objects can be used to represent highly general
+ * mathematical operations 1) as types, which can be passed around in template
```
Can you turn 1) and 2) into a Markdown list by putting them on a new line and doing `1.`?

nilsdeppe commented on src/Utilities/Functional.hpp:
````diff
@@ -7,11 +7,42 @@
+ *                          funcl::Plus<funlc::Identity, funcl::Multiplies<>>>>
+ * ```
+ * which then gives a function type which when instantiated and evaluated will
+ * give the answer \f$ a*sin(b + c *d)\f$ from calling Sinusoid{}(a,b,c,d)
````
`sin` to `\sin`

Ticks around sinusoid code

nilsdeppe commented on src/Utilities/Functional.hpp:
```diff
@@ -7,11 +7,42 @@
+ *
+ * Potentially useful as examples also are the generic tests from
+ * Test_Functional.cpp:
+ * \snippet Utilities/Test_Functional.cpp generic_function_test
```
I'm not sure the first snippet is as useful as it is overwhelming, but the second snippet even without the first is great for showing how the chaining works.

nilsdeppe commented on src/Utilities/Math.hpp:
```diff
@@ -45,3 +47,28 @@ T evaluate_polynomial(const std::vector<U>& coeffs, const T& x) noexcept {
       coeffs.rbegin(), coeffs.rend(), make_with_value<T>(x, 0.),
       [&x](const T& state, const U& element) { return state * x + element; });
 }
+
```
Could you:
- Factor these into a separate commit from where you add the functionals? (same PR, please)
- Add tests for them

nilsdeppe commented on tests/Unit/Utilities/Test_Functional.cpp:
```diff
@@ -3,16 +3,153 @@
+MAKE_UNARY_TEST(Sin, sin);
+MAKE_UNARY_TEST(Sinh, sinh);
+MAKE_UNARY_TEST(StepFunctionStruct, step_function);
+MAKE_UNARY_TEST(Sqrt, sqrt);
```
q before t

nilsdeppe commented on tests/Unit/Utilities/Test_Functional.cpp:
```diff
@@ -3,16 +3,153 @@
+MAKE_UNARY_TEST(Tan, tan);
+MAKE_UNARY_TEST(Tanh, tanh);
+
+MAKE_BINARY_TEST(Atan2, atan2);
```
did we lose the test for `Square`?

nilsdeppe commented on src/Utilities/Functional.hpp:
```diff
@@ -83,33 +195,49 @@ struct Identity : Functional<1> {
+MAKE_UNARY_FUNCTIONAL(Real, real);
+MAKE_UNARY_FUNCTIONAL(Sin, sin);
+MAKE_UNARY_FUNCTIONAL(Sinh, sinh);
+MAKE_UNARY_FUNCTIONAL(StepFunctionStruct, step_function);
```
Why not just `StepFunction`?

nilsdeppe commented on tests/Unit/Utilities/Test_Functional.cpp:
```diff
@@ -3,16 +3,153 @@
+MAKE_BINARY_INPLACE_TEST(MultAssign, *=, *);
+MAKE_BINARY_INPLACE_TEST(PlusAssign, +=, +);
+
+using RealTypeList = tmpl::list<float, double>;
```
Would testing `int` be as easy as adding it to this list? If so I'd be in favor of testing int and long as well. Same with complex<double> and int and long

moxcodes replied:
yep! I can just add to the list. Interesting point that I discovered doing so is that many functions just fail on `complex<integer types>`. I think we should just not try to support that for now.

nilsdeppe replied:
Okay, let's avoid them :) I personally don't know when we would even want a complex integer :/

nilsdeppe commented on tests/Unit/Utilities/Test_Functional.cpp:
```diff
@@ -22,56 +159,131 @@ void test_get_argument() {
-  CHECK(Plus<Identity, Multiplies<>>{}(1, 2, 3) == 7);
-}
+SPECTRE_TEST_CASE("Unit.Utilities.Functional", "[Utilities][Unit]") {
+  std::random_device r;
```
How many of the variables in this file can be const?

nilsdeppe commented on tests/Unit/Utilities/Test_Functional.cpp:
```diff
@@ -22,56 +159,131 @@ void test_get_argument() {
-}
+SPECTRE_TEST_CASE("Unit.Utilities.Functional", "[Utilities][Unit]") {
+  std::random_device r;
+  const auto seed = r();
```
Could you do an `INFO("Seed is: " << seed);` on the next line? This way if the test fails  we can reproduce it using the seed.

nilsdeppe commented on tests/Unit/Utilities/Test_Functional.cpp:
```diff
@@ -22,56 +159,131 @@ void test_get_argument() {
-  CHECK(Square<>{}(-3.0) == 9.0);
-}
 
-void test_composition() {
```
So the goal of `test_composition` was to make sure the basic design worked in terms of arity tracking, expecially for more complex examples like on 57, 58. It would be nice to retain that at least so that if `Functional` or something gets changed, we still know arity tracking is working.

nilsdeppe reviewed (changes requested):
I'm happy for you to squash the changes into the final commits you want to present :)

nilsdeppe commented on src/Utilities/Functional.hpp:
```diff
@@ -195,6 +223,11 @@ struct Identity : Functional<1> {
 };
 
+template <int val>
+struct Literal : Functional<0> {
```
What are the use cases for this that you're imagining? All this seems to do is replace an argument in `(...)` with one in the template list. (I'm not saying we don't want it, it's just not clear from this PR what the benefit of splitting numbers into two places is)

moxcodes replied:
The documentation I've added at the start of the namespace tries to clarify that - the usecase that I'm imagining is trying to assemble, for instance, a set of types which represent useful compile-time functions. For instance, say you want to create a set of one-dimensional probability density functions, which you'll want to all be Functional<1>'s. So, you want the ability to represent the full chain of ratios, products, etc that go on with compile-time literals in place. Once assembled, they could then be called on a template struct that does something with probability functions, like computes statistics without needing to know all the messy parameters.

nilsdeppe replied:
Okay, I can accept that :)

moxcodes commented:
fixed, tidied, and squished.

nilsdeppe commented on src/Utilities/Math.hpp:
```diff
@@ -45,3 +47,27 @@ T evaluate_polynomial(const std::vector<U>& coeffs, const T& x) noexcept {
+/// and complex types
+template <typename T, Requires<std::is_arithmetic<T>::value or
+                               tt::is_a_v<std::complex, T>> = nullptr>
+constexpr T invsqrt(const T& arg) noexcept {
```
These can't be `constexpr` until sqrt, cbrt, etc. are `constexpr`. It might be misleading to label them constexpr.

Note: On clang `__builtin_sqrt` does not appear to be `constexpr`, but it is in GCC

kidder replied:
ping

nilsdeppe commented on docs/Doxyfile.in:
```diff
@@ -156,7 +156,10 @@ ENABLE_PREPROCESSING   = YES
 
 # Macros we actually want to expand during preprocessing
-EXPAND_AS_DEFINED      = MAKE_BINARY_OPERATOR
+EXPAND_AS_DEFINED      = MAKE_BINARY_INPLACE_OPERATOR \
```
Alphabetize the macro names (I after F)

nilsdeppe commented on src/Utilities/Functional.hpp:
```diff
@@ -29,39 +79,128 @@ struct Functional {
+    }                                                                         \
+  } /** \endcond */
+
+#define MAKE_BINARY_FUNCTIONAL(NAME, OPERATOR)                                 \
```
F before I

nilsdeppe commented on src/Utilities/Functional.hpp:
```diff
@@ -83,33 +222,58 @@ struct Identity : Functional<1> {
 
-/// Functional for subtracting two objects
-MAKE_BINARY_OPERATOR(Minus, -)
+template <int val>
```
I think we should have a second template parameter `Type = double` so that it could be specified to use non-oduble types if desired

nilsdeppe commented on src/Utilities/Functional.hpp:
````diff
@@ -7,11 +7,61 @@
+ * instance, distributions. Let's make a gaussian with mean at 5.5 and a
+ * variance of .7.
+ * ```
+ * using FivePFive = funcl::Divides<funcl::Literal<11>,funcl::Literal<2>>;
````
Being overly verbose and writing `FivePointFive` would probably help people.

nilsdeppe commented on src/Utilities/Functional.hpp:
````diff
@@ -7,11 +7,61 @@
+ * using Gaussian = funcl::Multiplies<GaussianAmp,
+ *                       funcl::Exp<GaussianExp>>;
+ * ```
+ * This gives us a type which is a `Functional<1>` and has a call function that
````
`Functional` is not something people should know about. I'd prefer saying `This gives us a function object whose call operator takes one argument that gives`...

nilsdeppe commented on src/Utilities/Functional.hpp:
````diff
@@ -7,11 +7,61 @@
+ * ```
+ * This gives us a type which is a `Functional<1>` and has a call function that
+ * gives the value of the desired gaussian distribution \f$ \frac{1}{\sqrt{2 \pi
+ * .7}} \exp{- \frac{(x - 5.5)^2}{2 *.7}} \f$
````
This math rendered as latex looks a bit funny. Specifically the `.7`. Maybe `0.7` or 7 / 10?

nilsdeppe reviewed (commented):
I'm happy for you to squash these. The only thing I haven't thought enough about yet is the distributions. I'll think about it some more and then give you feedback.

moxcodes commented:
fixed and squashed direct changes, added new commit for changes to distribution and further related updates.
Cheers!

moxcodes commented on src/Utilities/Functional.hpp:
```diff
@@ -29,39 +79,128 @@ struct Functional {
+  template <class C0 = Identity, class C1 = C0>                                \
+  struct NAME : Functional<C0::arity + C1::arity> {                            \
+    using base = Functional<C0::arity + C1::arity>;                            \
+    template <class... Ts>                                                     \
```
Explore SFINAE with two variadic templates, maybe get rid of helper?

nilsdeppe replied:
SFINAE is extremely slow to compile, so in general I advise against it.

moxcodes replied:
Turns out greediness of param packs makes that not viable.

moxcodes commented on src/Utilities/Functional.hpp:
```diff
@@ -29,39 +79,128 @@ struct Functional {
+    }                                                                         \
+  };                                                                          \
+  /** \cond */                                                                \
+  template <class C1>                                                         \
```
maybe remove?

moxcodes commented on src/Utilities/Functional.hpp:
```diff
@@ -29,39 +79,128 @@ struct Functional {
+    }                                                     \
+  };                                                      \
+  template <>                                             \
+  struct NAME<Identity> : Functional<1> {                 \
```
maybe remove, add inline to identity

moxcodes commented on src/Utilities/Functional.hpp:
````diff
@@ -7,11 +7,61 @@
+ * template <class Sol, class... Ts> sample(Ts... params) { ...
+ * ```
+ * In the corresponding namespace, we could use these functional objects to
+ * define useful distirbutions, or generate them in the user code.
````
'distirbutions'

moxcodes commented on src/Utilities/Functional.hpp:
```diff
@@ -7,11 +7,61 @@
+ *
+ * Potentially useful as examples also are the generic tests from
+ * Test_Functional.cpp:
+ * \snippet Utilities/Test_Functional.cpp generic_function_test_invoke
```
add comments or variable name changes to clarify variables in this snippet

moxcodes commented on tests/Utilities/MakeWithRandomValues.hpp:
```diff
@@ -102,3 +156,24 @@ ReturnType make_with_random_values(
+/// `make_with_random_values<Type>(make_not_null(&gen),make_not_null(&dist)`
+template <typename T, typename UniformRandomBitGenerator,
+          typename RandomNumberDistribution>
+decltype(auto) make_with_random_values(
```
decltype(auto) -> T

moxcodes reviewed (commented):
group code review with @markscheel @mgiesler

moxcodes commented:
fixed and squashed again.
Cheers

nilsdeppe commented on src/Utilities/Functional.hpp:
```diff
@@ -93,22 +98,7 @@ struct Identity;
               std::make_index_sequence<C1::arity>{}));                         \
     }                                                                          \
-  };                                                                           \
-  /** \cond */                                                                 \
```
Please leave the `Identity` fast tracks.

nilsdeppe commented on src/Utilities/Functional.hpp:
```diff
@@ -217,7 +181,7 @@ struct GetArgument : Functional<Arity> {
 struct Identity : Functional<1> {
   template <class T>
-  constexpr const T& operator()(const T& t) noexcept {
+  inline constexpr const T& operator()(const T& t) noexcept {
```
Again, inline doesn't do what you think it does. `inline` is a statement about (roughly) linkage, not about whether or not the function is inlined. If you actually want the compiler to inline this we have `SPECTRE_ALWAYS_INLINE` in `Utilities/ForceInline.hpp`.

[Here](https://en.cppreference.com/w/cpp/language/inline) is what inline actually does.

nilsdeppe commented on src/Utilities/TypeTraits.hpp:
```diff
@@ -1360,6 +1360,30 @@ template <typename T>
+///
+/// `get_fundamental_type_t<T>` is provided as a type alias to `typename
+/// get_fundamental_type_t<T>::type`
+template <typename T, typename Enable = std::nullptr_t>
```
So what should happen if I have a container that does not have `::value_type`? If the answer is "It's undefined/ill-formed" then I think the correct way to implement this is with `void_t` rather than `Requires`. If it's not then this code is broken to begin with.

What about for containers like `unordered_map` where the `value_type` is a `pair`? How should that behave? Is that not a case covered?

moxcodes replied:
void_t sounds like a good solution for the second case, where I use a silly a==a test to make sure the value_type is there, but won't work for the first case, where I want to test for is_fundamental. I could use a `Requires<cpp:17::is_same_v<void_t<T::value_type>,void>>` for the second one, which is probably clearer.

Current behavior should be that if it can get to something that `is_fundamental_v<T>` via `::value_type`, the `type` in the struct is set to that, otherwise no type is set. `std::pair` doesn't have a `::value_type`, so an `unordered_map` would get no type field set. If this kind of usage is desired to return the `mapped_type`, I'd be happy to add more cases to the function.

nilsdeppe replied:
I have no issues saying this only works for `::value_type` containers (we can expand as necessary). In that case I would implement this as:

```cpp
template <typename T, typename cpp17::void_t<>>
struct get_fundamental_type {
  using type = T;
};
template <typename T>
struct get_fundamental_type<T, cpp17::void_t<typename T::value_type> {
  using type = typename get_fundamental_type<typename T::value_type>::type;
};
```
or, if you really want to do the fundamental check:
```cpp
template <typename T, typename cpp17::void_t<>>
struct get_fundamental_type {
  using type = tmpl::conditional_t<cpp17::is_fundamental_v<T>, T, NoSuchType>;
};
```

moxcodes replied:
nice! I hadn't known about conditional_t. That'll do nicely. :+1:

nilsdeppe replied:
Great! There's also a `std::conditional_t` but the `tmpl` one is slightly faster to compile :)

nilsdeppe commented on src/Utilities/TypeTraits.hpp:
```diff
@@ -1360,6 +1360,30 @@ template <typename T>
+/// \ingroup TypeTraitsGroup
+/// \brief Permits designating a fundamental type for a container
+///
+/// \details Uses SFINAE to designate `get_fundamental_type<T>::type` as `T`
```
Everything in type traits basically uses SFINAE, not something worth mentioning (if it's mentioned on other traits please file an issue and I'll remove it). The effects are interesting and should be documented :)

nilsdeppe commented on tests/Utilities/MakeWithRandomValues.hpp:
```diff
@@ -38,6 +37,23 @@ struct FillWithRandomValuesImpl<T, Requires<cpp17::is_fundamental_v<T>>> {
 };
 
+template <typename T>
+struct FillWithRandomValuesImpl<T, Requires<tt::is_a_v<std::complex, T>>> {
```
I believe the better way to do this is `FillWithRandomValuesImpl<std::complex<T>> {`. This avoids using SFINAE and just has the compiler pattern match, which is faster.

Note you'll need to write out `std::complex<T>` below, but that should be fine :)

nilsdeppe commented on tests/Utilities/MakeWithRandomValues.hpp:
```diff
@@ -73,67 +89,40 @@ struct FillWithRandomValuesImpl<Variables<tmpl::list<Tags...>>,
- private:
-  std::array<T, 2> bounds_;
-
+class UniformCustomDistribution<T, Requires<std::is_integral<T>::value>>
```
could you add to `Utilities/TypeTraits.hpp` `cpp17::is_integral_v`? Just follow the examples of `is_floating_point`. Thanks!

nilsdeppe commented on tests/Utilities/MakeWithRandomValues.hpp:
```diff
@@ -73,67 +89,40 @@ struct FillWithRandomValuesImpl<Variables<tmpl::list<Tags...>>,
- private:
-  std::array<T, 2> bounds_;
-
+class UniformCustomDistribution<T, Requires<std::is_floating_point<T>::value>>
```
Could you use `cpp17::is_floating_point_v<T>` from `Utilities/TypeTraits.hpp`?

nilsdeppe commented on tests/Utilities/MakeWithRandomValues.hpp:
```diff
@@ -73,67 +89,40 @@ struct FillWithRandomValuesImpl<Variables<tmpl::list<Tags...>>,
-                 std::real(gsl::at(bounds_, 0)));
-    return std::complex<T>(val1, val2);
-  }
+  explicit UniformCustomDistribution(std::array<Bound, 2> arr)
```
I would have `const std::array<T, 2>& arr` since the type must be convertible to `T` without precision loss, otherwise this is somewhat sketchy. Taking an array of `T` forces the person constructing the object to think about conversions instead of making an implicit choice for them.

moxcodes replied:
How about with the template but without casts? Otherwise, the array typing makes this constructor more strict than the base type constructor.

nilsdeppe replied:
Why would that be stricter? It would be the same as far as I can tell. If we have different definitions of the same, then I think what you want is `T{arr[0]}`, which is a safe conversion

moxcodes replied:
If not templated, an `array<double,2>` will not implicitly convert to an `array<int, 2>`, but doubles are fine in the constructor for `uniform_int_distribution`. I'd like this constructor to work for all arrays of types valid for the std constructor.

nilsdeppe replied:
Okay, then I think you want to do `T{arr[0]}, T{arr[1]}` to get the same behavior (compiler error if there's a precision losing conversion)

moxcodes replied:
hmm.. still stricter than the underlying constructor, which has no way of preventing implicit conversion. Doing this would ensure a lot of my general tests that use double arrays for generic bounds would just call the constructor with two arguments to get the easy implicit conversion, largely just skipping the convenience array constructor. If that's the desired style, though, it's only a small change in all the user code.

nilsdeppe commented on tests/Utilities/MakeWithRandomValues.hpp:
```diff
@@ -73,67 +89,40 @@ struct FillWithRandomValuesImpl<Variables<tmpl::list<Tags...>>,
-                static_cast<T>(Generator::max() - Generator::min()) +
-            gsl::at(bounds_, 0));
-  }
+  explicit UniformCustomDistribution(std::array<Bound, 2> arr)
```
See my comment below on `std::array<T, 2>`

nilsdeppe replied:
`noexcept` (same with constructor below`

nilsdeppe commented on tests/Utilities/MakeWithRandomValues.hpp:
```diff
@@ -167,3 +156,24 @@ ReturnType make_with_random_values(
+T make_with_random_values(
+    const gsl::not_null<UniformRandomBitGenerator*> generator,
+    const gsl::not_null<RandomNumberDistribution*> distribution) noexcept {
+  auto result = static_cast<T>(
```
Why do you need to do a cast here? I think what you want is for integral types to use their `max()` value instead, but I'm not sure.

moxcodes replied:
The cast is to optionally call the constructor of a fixed-size container type which can be created with a single value of the contained type, or just use the type itself if it is fundamental. If the cast is insufficiently safe, I can make a checker for `TypeTraits.hpp` to verify that the object can be so constructed and `static_assert` before this line

nilsdeppe replied:
Then I think you should do a safe conversion/construction using `T result{...};` (braces are important)

nilsdeppe commented on tests/Utilities/MakeWithRandomValues.hpp:
```diff
@@ -73,67 +89,40 @@ struct FillWithRandomValuesImpl<Variables<tmpl::list<Tags...>>,
+/// point or int distributions
+// {@
+template <typename T, typename Enable = std::nullptr_t>
+class UniformCustomDistribution {};
```
remove ` {}` since this should only be a forward declaration.

kidder commented on src/Utilities/Functional.hpp:
```diff
@@ -9,12 +9,67 @@
+ * comparing the result of its template argument with the function in its first
+ * function argument on random values. The third argument is an
+ * `array<double,2>` of bounds for the distribution, and the third argument is
+ * an index sequence of lenght equal to the number of arguments in the supplied
```
length

kidder commented on src/Utilities/Functional.hpp:
```diff
@@ -9,12 +9,67 @@
+ *  \snippet Utilities/Test_Functional.cpp generic_function_test_invoke
  */
 namespace funcl {
+using std::conj;
```
comment that using these for overload resolution with blaze

fmahebert commented on src/Utilities/Functional.hpp:
```diff
@@ -9,12 +9,67 @@
+ * 1. as types, which can be passed around in template arguments, and
+ * 2. such that any time they can be evaluated at compile time, they will be.
+ *
+ * As an example usecase, consider a function templated on a generic analytic
```
usecase -> use case

fmahebert commented on src/Utilities/Functional.hpp:
````diff
@@ -9,12 +9,67 @@
+ * In the corresponding namespace, we could use these functional objects to
+ * define useful distributions, or generate them in the user code.
+ * ```
+ * using Sinusoid = funcl::Multiplies<funcl::Identity,funcl::Sin<
````
Add space after comma for readability :)

Also, can you turn this into a doxygen snippet?

fmahebert commented on src/Utilities/Functional.hpp:
````diff
@@ -9,12 +9,67 @@
+ * using Sinusoid = funcl::Multiplies<funcl::Identity,funcl::Sin<
+ *                          funcl::Plus<funcl::Identity, funcl::Multiplies<>>>>
+ * ```
+ * which then gives a function type which when instantiated and evaluated will
````
Remove word "function"

fmahebert commented on src/Utilities/Functional.hpp:
````diff
@@ -9,12 +9,67 @@
+ *                          funcl::Plus<funcl::Identity, funcl::Multiplies<>>>>
+ * ```
+ * which then gives a function type which when instantiated and evaluated will
+ * give the answer \f$ a*\sin(b + c *d)\f$ from calling `Sinusoid{}(a,b,c,d)`
````
In the latex, use `\times`

fmahebert commented on src/Utilities/Functional.hpp:
````diff
@@ -9,12 +9,67 @@
+ * solution
+ * ```
+ * template <class Sol, class... Ts> sample(Ts... params) { ...
+ * ```
````
I don't see the above example being used anywhere in the remainder of the documentation. Either remove this, or refer back to it from somewhere below.

fmahebert commented on src/Utilities/Functional.hpp:
```diff
@@ -9,12 +9,67 @@
+ * which then gives a function type which when instantiated and evaluated will
+ * give the answer \f$ a*\sin(b + c *d)\f$ from calling `Sinusoid{}(a,b,c,d)`
+ *
+ * As a more creative example, we can take advantage of literals to make, for
```
The two examples below are slightly overwhelming due to their length and complexity, and they are unlikely to represent a realistic use case. We think the sinusoid example is sufficient and the below should be dropped from this documentation.

moxcodes replied:
I'm happy to boil the gaussian example down to a more manageable level, but as a significant motivation in creating these was ease in making general tests, I think at least some exposure to the below tests is valuable.

kidder replied:
so for the final example you had there was no relation between the documentation and the code snippet

moxcodes replied:
Sorry, I don't understand; Did the code snippet come out wrong? It's intended to be a bunch of function calls to a generic function testing a functor type against a lambda, which is intended to be described immediately prior. I had previously included the full function definition previously as well, but @nilsdeppe suggested just the calls were more informative than the function definition as well.

kidder replied:
the rendered documentation described testing with random values and the snippet appeared as a bunch of calls to a function with no relation to that description.

moxcodes replied:
thanks, understood. The documentation was intended to be a description of the function definition now omitted to provide context. I'll trim the calls and rewrite the documentation so it is clearer.

kidder commented on src/Utilities/Functional.hpp:
```diff
@@ -78,8 +102,128 @@ struct AssertEqual : Functional<2> {
+  } /** \endcond */
+
+#define MAKE_BINARY_INPLACE_OPERATOR(NAME, OPERATOR)                          \
+  /** Functional for computing `OPERATOR` of two objects*/                    \
```
`objects */`

fmahebert reviewed (changes requested):
doxygen reviewed with @wthrowe @nilsdeppe @kidder

kidder commented on src/Utilities/Functional.hpp:
```diff
@@ -78,8 +102,128 @@ struct AssertEqual : Functional<2> {
+    using base = Functional<C0::arity + C1::arity>;                           \
+    template <class... Ts>                                                    \
+    constexpr auto operator()(Ts&... ts) noexcept {                           \
+      return (                                                                \
```
remove outer `()`

kidder commented on src/Utilities/Functional.hpp:
```diff
@@ -78,8 +102,128 @@ struct AssertEqual : Functional<2> {
+  template <class C1>                                                         \
+  struct NAME<Identity, C1> : Functional<1 + C1::arity> {                     \
+    template <class T0, class... Ts>                                          \
+    constexpr auto operator()(T0& t0, const Ts&... ts) noexcept {             \
```
return `decltype(auto)`

kidder commented on src/Utilities/Functional.hpp:
```diff
@@ -78,8 +102,128 @@ struct AssertEqual : Functional<2> {
+  struct NAME : Functional<C0::arity + C1::arity> {                           \
+    using base = Functional<C0::arity + C1::arity>;                           \
+    template <class... Ts>                                                    \
+    constexpr auto operator()(Ts&... ts) noexcept {                           \
```
return `decltype(auto)`

kidder commented on src/Utilities/Functional.hpp:
```diff
@@ -78,8 +102,128 @@ struct AssertEqual : Functional<2> {
+  struct NAME<Identity, Identity> : Functional<2> {                           \
+    static constexpr size_t arity = 2;                                        \
+    template <class T0, class T1>                                             \
+    constexpr auto operator()(T0& t0, const T1& t1) noexcept {                \
```
return `decltype(auto)`

kidder commented on src/Utilities/Functional.hpp:
```diff
@@ -78,8 +102,128 @@ struct AssertEqual : Functional<2> {
+  } /** \endcond */
+
+#define MAKE_BINARY_OPERATOR(NAME, OPERATOR)                                  \
+  /** Functional for computing `OPERATOR` of two objects*/                    \
```
`objects */`

kidder commented on src/Utilities/Functional.hpp:
```diff
@@ -78,8 +102,128 @@ struct AssertEqual : Functional<2> {
+
+#define MAKE_LITERAL_VAL(NAME, VAL)          \
+  /** Functional literal for `VAL` */        \
+  struct Lit##NAME : Functional<0> {         \
```
Lit -> Literal

kidder commented on src/Utilities/Functional.hpp:
```diff
@@ -96,38 +240,63 @@ struct GetArgument : Functional<Arity> {
-MAKE_BINARY_OPERATOR(Minus, -)
+template <int val, typename Type = double>
+struct Literal : Functional<0> {
+  constexpr double operator()() noexcept { return static_cast<Type>(val); }
```
`double -> Type`

kidder commented on src/Utilities/Math.hpp:
```diff
@@ -45,3 +47,27 @@ T evaluate_polynomial(const std::vector<U>& coeffs, const T& x) noexcept {
 }
+
+/// \ingroup UtilitiesGroup
+/// \brief defines the step function for arithmetic types
```
Defines.

Specify what 0 is

kidder commented on src/Utilities/Math.hpp:
```diff
@@ -45,3 +47,27 @@ T evaluate_polynomial(const std::vector<U>& coeffs, const T& x) noexcept {
+/// \brief defines the step function for arithmetic types
+template <typename T, Requires<std::is_arithmetic<T>::value> = nullptr>
+constexpr T step_function(const T& arg) noexcept {
+  return (arg > 0) ? 1 : 0;
```
`static_cast<T> ((arg > static_cast<T>(0) ) ? 1 : 0)`

kidder commented on src/Utilities/TypeTraits.hpp:
```diff
@@ -1360,6 +1365,31 @@ template <typename T>
+struct get_fundamental_type {
+  using type = tmpl::conditional_t<cpp17::is_fundamental_v<T>, T, NoSuchType>;
+};
+template <typename T>
```
`/// \cond`

kidder commented on src/Utilities/TypeTraits.hpp:
```diff
@@ -1360,6 +1365,31 @@ template <typename T>
 using remove_cvref_wrap_t = typename remove_cvref_wrap<T>::type;
 // @}
 
+/// \ingroup TypeTraitsGroup
```
`// @{`

kidder replied:
render and fix

kidder commented on src/Utilities/TypeTraits.hpp:
```diff
@@ -1360,6 +1365,31 @@ template <typename T>
+};
+template <typename T>
+using get_fundamental_type_t = typename get_fundamental_type<T>::type;
+
```
`// @}`

kidder commented on src/Utilities/TypeTraits.hpp:
```diff
@@ -1360,6 +1365,31 @@ template <typename T>
+struct get_fundamental_type<T, cpp17::void_t<typename T::value_type>> {
+  using type = typename get_fundamental_type<typename T::value_type>::type;
+};
+template <typename T>
```
`/// \endcond`

kidder commented on tests/Unit/Utilities/Test_TypeTraits.cpp:
```diff
@@ -693,6 +694,18 @@ static_assert(
     "Failed testing remove_cvref_wrap");
 /// [remove_cvref_wrap]
 
+/// [get_fundamental_type]
```
also test that if you pass a fundamental you get a fundamental

kidder commented on tests/Unit/Utilities/Test_Math.cpp:
```diff
@@ -30,4 +30,14 @@ SPECTRE_TEST_CASE("Unit.Utilities.Math", "[Unit][Utilities]") {
+  SECTION("Test inverse roots and step_function") {
+    CHECK(step_function(1.0) == 1.0);
+    CHECK(step_function(0.5) == 1.0);
+    CHECK(step_function(-10) == 0);
```
test at zero (with ints and doubles)

kidder commented on tests/Unit/Utilities/Test_Math.cpp:
```diff
@@ -30,4 +30,14 @@ SPECTRE_TEST_CASE("Unit.Utilities.Math", "[Unit][Utilities]") {
+    CHECK(step_function(-10) == 0);
+    CHECK(invsqrt(4.0) == 0.5);
+    CHECK(invsqrt(10) == 0);
+    CHECK_ITERABLE_APPROX(invcbrt(27.0), (1 / 3.0));
```
CHECK(approx(a)==b);

moxcodes commented:
New fixups posted.
Cheers!

nilsdeppe commented on src/Utilities/TypeTraits.hpp:
```diff
@@ -1365,30 +1365,29 @@ template <typename T>
 };
 template <typename T>
 using get_fundamental_type_t = typename get_fundamental_type<T>::type;
+/// \endcond
```
This should be *before* the type alias

nilsdeppe commented on tests/Unit/Utilities/Test_Functional.cpp:
```diff
@@ -181,7 +180,19 @@ void test_functional_against_function(
 
-void test_get_argument() {
+/// [using_sinusoid]
+using Sinusoid = funcl::Multiplies<
```
I might just not be seeing it, but these type aliases should be tested to make sure they evaluate to the right thing.

nilsdeppe commented on tests/Unit/Utilities/Test_Math.cpp:
```diff
@@ -35,9 +35,11 @@ SPECTRE_TEST_CASE("Unit.Utilities.Math", "[Unit][Utilities]") {
+    CHECK(step_function(0.0) == 1.0);
+    CHECK(step_function(0) == 1);
     CHECK(invsqrt(4.0) == 0.5);
     CHECK(invsqrt(10) == 0);
```
This is the wrong answer, should be `1 / sqrt(10)` but is zero.

nilsdeppe commented on src/Utilities/Functional.hpp:
```diff
@@ -121,115 +94,117 @@ struct AssertEqual : Functional<2> {
+  struct NAME : Functional<C0::arity + C1::arity> {                            \
+    using base = Functional<C0::arity + C1::arity>;                            \
+    template <class... Ts>                                                     \
+    constexpr decltype(auto) operator()(const Ts&... ts) noexcept {            \
```
Nope, we only asked for `decltype(auto)` on the inplace one.

nilsdeppe commented on src/Utilities/Functional.hpp:
```diff
@@ -121,115 +94,117 @@ struct AssertEqual : Functional<2> {
 struct GetArgument : Functional<Arity> {
   template <class... Ts>
-  constexpr auto operator()(const Ts&... ts) noexcept {
+  constexpr decltype(auto) operator()(const Ts&... ts) noexcept {
```
This is fine as `decltype(auto)` :+1:

nilsdeppe commented on src/Utilities/TypeTraits.hpp:
```diff
@@ -1365,30 +1365,29 @@ template <typename T>
+// @{
 /// \ingroup TypeTraitsGroup
-/// \brief Permits designating a fundamental type for a container
+/// \brief Extracts a fundamental type for a container
```
`a funda...` to `the funda...`?

nilsdeppe commented on src/Utilities/Functional.hpp:
```diff
@@ -22,50 +22,23 @@
- * which then gives a function type which when instantiated and evaluated will
- * give the answer \f$ a*\sin(b + c *d)\f$ from calling `Sinusoid{}(a,b,c,d)`
+ * As an illustrative example, consider the definition of a general sinusoid
+ * functor type :
```
`functor` -> `function object`

moxcodes commented:
new mistakes fixed. Thanks!

nilsdeppe reviewed (commented):
I'm fine for this to be squashed and factored into a few logical commits (single PR), but wait for @kidder to take a look at the fixes.

kidder commented:
I am happy with the fixes

moxcodes committed:
Add math functions invsqrt, invcbrt, and step_function for primitives

moxcodes committed:
add `get_fundamental_type`

- also added utility type alias `is_integral_v`

moxcodes committed:
add functors for standard and blaze math functions

Put into Utilities/Functional.hpp are macros for unary, binary, and binary
assign operators such that functors can easily be generated from a function
argument. This is then used to define functors associated with all supported
operations on DataVectors. For convenience in writing type-generic code,
step_function, invsqrt, and invcbrt are defined for primitives.

Also added the functionality for literal functionals for ease in creating more
intricate functionals, as well as increased documentation for functionals.

moxcodes committed:
Add convenience class for Uniform distribution of arbitrary type

also:
- added `FillWithRandomValuesImpl` for complex
- added convenience function for make_with_random_values on fixed-length types

moxcodes commented:
Great, I've rewritten history to be prettier now.
Cheers!

nilsdeppe commented on tests/Utilities/MakeWithRandomValues.hpp:
```diff
@@ -70,6 +87,40 @@ struct FillWithRandomValuesImpl<Variables<tmpl::list<Tags...>>,
+///
+/// \brief A uniform distribution function object which redirects appropriately
+/// to either the `std::uniform_int_distribution` or the
+/// `std::uniform_real_distribution` as appropriate. also provides a convenience
```
Remove `as appropriate` (you already said this earlier in the sentence)

`also` -> `Also`

nilsdeppe commented on tests/Utilities/MakeWithRandomValues.hpp:
```diff
@@ -70,6 +87,40 @@ struct FillWithRandomValuesImpl<Variables<tmpl::list<Tags...>>,
+/// \brief A uniform distribution function object which redirects appropriately
+/// to either the `std::uniform_int_distribution` or the
+/// `std::uniform_real_distribution` as appropriate. also provides a convenience
+/// constructor which takes a 2-element array for the bounds.  either floating
```
why the period and double space between `bounds` and `either floating`?

nilsdeppe commented on tests/Utilities/MakeWithRandomValues.hpp:
```diff
@@ -102,3 +153,39 @@ ReturnType make_with_random_values(
+/// type, fills it with random values, and returns the result. Acts as a
+/// convenience function to avoid users needing to put in constructors with
+/// `signaling_NaN()`s or `max()`s themselves when making with random values.
+///
```
spacing?

nilsdeppe commented on tests/Unit/Utilities/Test_Functional.cpp:
```diff
@@ -3,21 +3,196 @@
+
+#define MAKE_BINARY_INPLACE_TEST(STRUCTNAME, OP, TESTOP)                       \
+  struct TestFuncEval##STRUCTNAME {                                            \
+    template <typename T1, typename T2, typename DistT1, typename DistT2,      \
```
I would find a comment explaining why the overloads exist instead of just one helpful :)

nilsdeppe commented on tests/Unit/Utilities/Test_Functional.cpp:
```diff
@@ -26,57 +201,169 @@ void test_get_argument() {
+  tmpl::for_each<AllTypeList>([&generator, &generic_unaries ](auto x) noexcept {
+    using DistType =
+        typename tt::get_fundamental_type_t<typename decltype(x)::type>;
+    tuple_fold(generic_unaries, [&generator, &x ](auto tup) noexcept {
```
`auto` -> `const auto&` (no point in making a copy each time)

nilsdeppe commented on tests/Unit/Utilities/Test_Functional.cpp:
```diff
@@ -26,57 +201,169 @@ void test_get_argument() {
+  ](auto x) noexcept {
+    using DistType1 =
+        typename tt::get_fundamental_type_t<typename decltype(x)::type>;
+    tuple_fold(floating_unaries, [&generator, &x ](auto tup) noexcept {
```
`auto` -> `const auto&` (no point in making a copy each time)

moxcodes replied:
happy to add reference, but const causes problems:
it means the call operator in the macro-generated functions have to be const, which means the distributions passed in have to be const, which causes compiler errors at the point of random number generation because std::uniform...distributions(generator) are not const.

moxcodes replied:
oops, nope - references cause the same problem because `tuple_fold_impl` consts its tuples too; so, I either need to make copies or change the way the functions work. For these tests, the copies are pretty small (empty function object and two doubles) - so I think it should be fine. I guess it is conceivable that having a tuple_fold that doesn't const the tuple is desirable, but seems like a discussion beyond this PR.

nilsdeppe commented on tests/Unit/Utilities/Test_Functional.cpp:
```diff
@@ -26,57 +201,169 @@ void test_get_argument() {
+              typename tt::get_fundamental_type_t<typename decltype(y)::type>;
+          tuple_fold(
+              tuple_cat(generic_binaries, floating_binaries),
+              [&generator, &x, &y ](auto tup) noexcept {
```
`auto` -> `const auto&` (no point in copying each time)

nilsdeppe commented on tests/Unit/Utilities/Test_Functional.cpp:
```diff
@@ -26,57 +201,169 @@ void test_get_argument() {
+  std::random_device r;
+  const auto seed = r();
+  INFO("Seed (individual functionals) is: " << seed);
+  std::mt19937 generator(seed);
```
I would find this much easier to read if you created functions like `void test_generic_unaries` and then called those. That way I could easily see where a thought starts and ends. It also means these things will likely follow the "All functions must fit in a single column vertically on a screen, otherwise it's not a small enough unit" rule

nilsdeppe commented on tests/Unit/Utilities/Test_Functional.cpp:
```diff
@@ -26,57 +201,169 @@ void test_get_argument() {
+        });
+      });
+  test_functional_against_function<Plus<Minus<>, Identity>, double>(
+      [](auto x, auto y, auto z) { return (x - y) + z; }, generic,
```
`auto` -> `const auto&` in these lambda expressions

nilsdeppe reviewed (changes requested):
I'm happy for you to squash

Note: I haven't rendered the docs.

moxcodes commented:
new test updates squashed in.
Cheers!

nilsdeppe commented on tests/Unit/Utilities/Test_Functional.cpp:
```diff
@@ -26,57 +201,202 @@ void test_get_argument() {
+          typename tt::get_fundamental_type_t<typename decltype(y)::type>;
+      tuple_fold(std::tuple_cat(real_binaries, generic_binaries), [
+        &gen, &x, &y
+      ](auto tup) noexcept {
```
After you mentioned the `const &` not working I realized that's because you are not doing a fold, you are doing a `for_each` with a fold. You should just do:
```
tmpl::for_each<tmpl::make_sequence<std::integral_constant<size_t, 0>, 4, tmpl::next<tmpl::_1>>>(
  [&gen, &x, &y, &tuple](auto index) noexcept {
    auto& tup = std::get<declytpe(index)::value>(tuple);
    std::get<0>(tup)(...);
  });
```

nilsdeppe reviewed (changes requested):
One last thing from me. Again I'm happy for you to squash :+1: nice job!

moxcodes commented:
Another squash completed.
Cheers!

mar-celine commented on tests/Utilities/MakeWithRandomValues.hpp:
```diff
@@ -102,3 +153,38 @@ ReturnType make_with_random_values(
+/// type, fills it with random values, and returns the result. Acts as a
+/// convenience function to avoid users needing to put in constructors with
+/// `signaling_NaN()`s or `max()`s themselves when making with random values.
+/// used as
```
`Used as`

nilsdeppe commented on tests/Utilities/MakeWithRandomValues.hpp:
```diff
@@ -70,6 +87,40 @@ struct FillWithRandomValuesImpl<Variables<tmpl::list<Tags...>>,
+/// `std::uniform_real_distribution`. This also provides a convenience
+/// constructor which takes a 2-element array for the bounds for either
+/// floating point or int distributions.
+template <typename T, typename Enable = std::nullptr_t>
```
During group code review (with tmpl and cpp17 swapped in for conditional_t and ::value):
```cpp
template <typename T>
class UniformCustomDistribution
    : public std::conditional_t<std::is_integral<T>::value,
                                std::uniform_int_distribution<T>,
                                std::uniform_real_distribution<T>> {
  using base = std::conditional_t<std::is_integral<T>::value,
                                  std::uniform_int_distribution<T>,
                                  std::uniform_real_distribution<T>>;
  static_assert(std::is_integral<T>::value or std::is_floating_point<T>::value,
                "Uh oh");

 public:
  using base::base;
  template <typename Bound>
  explicit UniformCustomDistribution(std::array<Bound, 2> arr) noexcept
      : base(arr[0], arr[1]) {}
  using base::operator=;
  using base::operator();
};
```

mar-celine commented on src/Utilities/Functional.hpp:
```diff
@@ -32,7 +60,86 @@ struct Functional {
 /// \endcond
 
+/// Functional that asserts the first and second arguments are equal and returns
+/// the first argument.
```
`returns C applied to the first argument`and we also want to have `C{}(t0)` and `C{}(t1)`s equality checked, not t0 and t1

mar-celine commented on tests/Unit/Utilities/Test_Functional.cpp:
```diff
@@ -3,21 +3,195 @@
+MAKE_BINARY_INPLACE_TEST(MultAssign, *=, *);
+MAKE_BINARY_INPLACE_TEST(PlusAssign, +=, +);
+
+using RealTypeList = tmpl::list<float, double, int long>;
```
Did you mean `int, long`?

nilsdeppe reviewed (changes requested):
squash from group code review (including what @gsb76 said). People present:

@gsb76 @kidder @wthrowe @nilsdeppe

moxcodes commented:
fixed and squashed.
Cheers!
